File: 2025/6/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"strconv"
	"strings"

	"github.com/weitecklee/adventofcode/utils"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	numbers, ops := parseInput(strings.Split(string(data), "\n"))
	fmt.Println(part1(numbers, ops))
	fmt.Println(part2(strings.Split(string(data), "\n")))
}

func parseInput(data []string) ([][]int, string) {
	numbers := make([][]int, len(data)-1)
	for i := range len(data) - 1 {
		numbers[i] = utils.ExtractInts(data[i])
	}
	line := data[len(data)-1]
	spaceRegex := regexp.MustCompile(`\s+`)
	line = spaceRegex.ReplaceAllString(line, "")
	return numbers, line
}

func part1(numbers [][]int, ops string) int {
	res := 0
	for col := range len(numbers[0]) {
		var curr int
		if ops[col] == '+' {
			for row := range numbers {
				curr += numbers[row][col]
			}
		} else {
			curr = 1
			for row := range numbers {
				curr *= numbers[row][col]
			}
		}
		res += curr
	}
	return res
}

func part2(puzzleInput []string) int {
	res := 0
	ops := puzzleInput[len(puzzleInput)-1]
	spaceRegex := regexp.MustCompile(`\s+`)
	ops = spaceRegex.ReplaceAllString(ops, "")
	puzzleInput = puzzleInput[:len(puzzleInput)-1]
	rowLen := len(puzzleInput[0])
	for _, row := range puzzleInput {
		if len(row) > rowLen {
			rowLen = len(row)
		}
	}
	transposed := make([]string, rowLen)
	for i := range rowLen {
		var s strings.Builder
		for _, line := range puzzleInput {
			if i >= len(line) {
				s.WriteByte(' ')
			} else {
				s.WriteByte(line[i])
			}
		}
		transposed[i] = s.String()
	}
	var numbers []int
	var j int
	for _, s := range transposed {
		n, err := strconv.Atoi(strings.TrimSpace(s))
		if err != nil {
			var tmp int
			if ops[j] == '+' {
				for _, n := range numbers {
					tmp += n
				}
			} else {
				tmp = 1
				for _, n := range numbers {
					tmp *= n
				}
			}
			res += tmp
			numbers = make([]int, 0)
			j += 1
		} else {
			numbers = append(numbers, n)
		}
	}
	var tmp int
	if ops[j] == '+' {
		for _, n := range numbers {
			tmp += n
		}
	} else {
		tmp = 1
		for _, n := range numbers {
			tmp *= n
		}
	}
	res += tmp
	return res
}

File: 2025/6/solution.ts
import * as fs from "fs";
import * as path from "path";

const puzzleInput = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split("\n");

const numRegex = /\d+/g;
const opRegex = /[\+\*]/g;

function parseInput(data: string[]): [number[][], string[]] {
  const opsString = data[data.length - 1];
  const ops = opsString.match(opRegex)!;

  const numbers: number[][] = [];
  for (const line of data.slice(0, -1)) {
    numbers.push(line.match(numRegex)!.map(Number));
  }

  return [numbers, ops];
}

function part1(numbers: number[][], ops: string[]): number {
  let res = 0;
  for (let i = 0; i < ops.length; i++) {
    let curr = 0;
    if (ops[i] == "*") {
      curr = 1;
      for (const nums of numbers) {
        curr *= nums[i];
      }
    } else {
      for (const nums of numbers) {
        curr += nums[i];
      }
    }
    res += curr;
  }
  return res;
}

function part2(puzzleInput: string[]): number {
  let opsString = puzzleInput[puzzleInput.length - 1];
  const numStrings = puzzleInput.slice(0, -1);

  let maxLen = 0;
  for (const line of numStrings) {
    maxLen = Math.max(maxLen, line.length);
  }

  for (let i = 0; i < numStrings.length; i++) {
    numStrings[i] = numStrings[i].padEnd(maxLen, " ");
  }
  opsString = opsString.padEnd(maxLen, " ");

  let res = 0;
  let currNums = [];
  for (let i = maxLen - 1; i >= 0; i--) {
    let curr = "";
    for (let j = 0; j < numStrings.length; j++) {
      curr += numStrings[j][i];
    }
    currNums.push(Number(curr.trim()));
    if (opsString[i] !== " ") {
      if (opsString[i] === "*") {
        res += currNums.reduce((a, b) => a * b, 1);
      } else if (opsString[i] === "+") {
        res += currNums.reduce((a, b) => a + b);
      }
      currNums = [];
      i--;
    }
  }

  return res;
}

const [numbers, ops] = parseInput(puzzleInput);
console.log(part1(numbers, ops));
console.log(part2(puzzleInput));
